server/CGameHandler.cpp

```diff
@@ -2233,6 +2242,53 @@ bool CGameHandler::razeStructure (ObjectInstanceID tid, BuildingID bid)
 	return true;
 }
 
+bool CGameHandler::spellResearch(ObjectInstanceID tid, SpellID spellAtSlot)
+{
+	if(!getSettings().getBoolean(EGameSettings::TOWNS_SPELL_RESEARCH) && complain("Spell research not allowed!"))
+		return false;
+
+	CGTownInstance *t = gs->getTown(tid);
+
+	int level = -1;
+	for(int i = 0; i < t->spells.size(); i++)
+		if(vstd::find_pos(t->spells[i], spellAtSlot) != -1)
+			level = i;
+	
+	if(level == -1 && complain("Spell for replacement not found!"))
+		return false;
+	
+	int daysSinceLastResearch = gs->getDate(Date::DAY) - t->lastSpellResearchDay;
+	if(!daysSinceLastResearch && complain("Already researched today!"))
+		return false;
+
+	TResources cost;
+	cost[EGameResID::GOLD] = 1000;
+	cost[EGameResID::MERCURY] = (level + 1) * 2;
+	cost[EGameResID::SULFUR] = (level + 1) * 2;
+	cost[EGameResID::CRYSTAL] = (level + 1) * 2;
+	cost[EGameResID::GEMS] = (level + 1) * 2;
+
+	if(!getPlayerState(t->getOwner())->resources.canAfford(cost) && complain("Spell replacement cannot be afforded!"))
+		return false;
+
+	giveResources(t->getOwner(), -cost);
+
+	auto spells = t->spells.at(level);
+
+	std::swap(spells.at(t->spellsAtLevel(level, false)), spells.at(vstd::find_pos(spells, spellAtSlot)));
+	auto it = spells.begin() + t->spellsAtLevel(level, false);
+	std::rotate(it, it + 1, spells.end()); // move to end
```

Comment:
This means that once you roll through all spells you can learn *all* of them.
AFAIK hota spell research allows you to only learn spell once you've "finished" spell research.
So if you're researching lvl 5 spell your hero can only learn spell that you had originally and final spell that you've confirmed.

IMO if we want simple version of spell research something like this is better option:
> when player requests a spell reroll, game will propose replacement spell that player may accept or deny
> - if player accepts, spell will be added to list of obligatory spells
> - if player denies, spell will be added to list of banned spell (yes, permanently)
> - if player denies, he can reroll spell again (unless there is daily limit)
> - if all spells are either obligatory or banned, spell reroll is not possible

Reply:
seems to be better, but i see abuse potential because if it works the way that I think then all spell slots of same tier share banned list then player may use first spell slots as "always accept if spell is at least mediocre" kind of thing and only resort to use last spell slot for getting the spell he wants

Reply:
changed the algorithm a little bit:
- it's now possible to skip spell (can do a new decision next day)
- each accepted research the costs increases exponential. Adds some strategic decisions to make. Avoids also to learn all (at least without unlimited resources).